cloudstack_plugin/volume.py

```diff
@@ -0,0 +1,104 @@
+########
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# * See the License for the specific language governing permissions and
+# * limitations under the License.
+
+
+from cloudify import ctx
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+
+from cloudstack_plugin.cloudstack_common import (
+    get_cloud_driver,
+    get_resource_id,
+    CLOUDSTACK_ID_PROPERTY,
+    CLOUDSTACK_NAME_PROPERTY,
+    CLOUDSTACK_TYPE_PROPERTY,
+    COMMON_RUNTIME_PROPERTIES_KEYS
+
+)
+from cloudstack_plugin.virtual_machine import get_vm_by_id
+
+VOLUME_CLOUDSTACK_TYPE = 'volume'
+
+RUNTIME_PROPERTIES_KEYS = COMMON_RUNTIME_PROPERTIES_KEYS
+
+
+@operation
+def attach_volume(**kwargs):
+    """ Create volume and attach to virtual machine.
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    vm_id = ctx.target.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    vm = get_vm_by_id(ctx, cloud_driver, vm_id)
+
+    volume_name = ctx.source.node.properties['name']
+    volume_size = ctx.source.node.properties['size']
+
+    volume = cloud_driver.create_volume(name=volume_name,
+                                        size=volume_size)
+
+    cloud_driver.attach_volume(node=vm,
+                               volume=volume)
+
+    ctx.source.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+    ctx.source.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+        VOLUME_CLOUDSTACK_TYPE
```

Comment:
not sure i follow. what is deviceid 1 and deviceid 2? why are there two of them?

there are two options for the built in file system type to work:
1. the volume node contains a 'device_name' property that is the actual device name. This is good in case the API accepts such an argument and can guaranty that this device name is correct. since, like you said, this highly depends on the hypervisor, most cloud providers cannot enforce it.
2. option two, is that the attach operation will set a runtime property on the source called 'device_name'.

so given that, is there any way of achieving it with cloudstack? 
for example, openstack has an option to pass 'auto' as the device name, and later you can query the attachement and it will return the actual device name.

sorry for the long comment, just wanted to explain the missing parts.